### PR TITLE
Add autofixer to `block-indentation` rule

### PR DESCRIPTION
**Pull request by VincentMolinie**

### CODE REVIEW

**Comment by bmish** [comment]
Hope to see this and #2271 land!

**Comment by VincentMolinie** [comment]
> Hope to see this and #2271 land!

Yeah this one seems good I think. `attribute-indetation` needs some reqorks as I'm actually fixing `block-identation` inside it 😅

**Comment by VincentMolinie** [comment]
tests pass @bmish 🙌

**Comment by bmish** [comment]
Looks like you need to run `yarn update:readme`.

File: lib/rules/block-indentation.js [hunk 1]
```diff
@@ -21,7 +21,7 @@
 
  <div>
   <p>{{t "greeting"}}</p>
- </div>
+  </div>
```
Comment (bmish) [inline lib/rules/block-indentation.js:24]: This looks wrong?

File: lib/rules/block-indentation.js [hunk 2]
```diff
@@ -161,17 +175,33 @@ module.exports = class BlockIndentation extends Rule {
...
   visitor() {
     this._elementStack = [];
+    this.seen = new Set();
+    this.nodeToSource = new WeakMap();
 
     return {
+      Template(node) {
+        if (this.mode === 'fix') {
+          if (
+            node.body[0] &&
```
Comment (bmish) [inline lib/rules/block-indentation.js:185]: Better way to write this is: `node.body.length > 0`

File: lib/rules/block-indentation.js [hunk 3]
```diff
@@ -161,17 +175,33 @@ module.exports = class BlockIndentation extends Rule {
...
+            node.body[0].type === 'TextNode' &&
+            !node.body[0].chars.replace(/ +/, '')
```
Comment (bmish) [inline lib/rules/block-indentation.js:187]: Can you extract a helper function for this `isFoo(str)` to make it more clear what this is testing?

File: lib/rules/block-indentation.js [hunk 4]
```diff
@@ -81,6 +83,18 @@ function isControlChar(char) {
   return char === '~' || char === '{' || char === '}';
 }
 
+function get(node, path) {
```
Comment (bmish) [inline lib/rules/block-indentation.js:86]: Can you add a jsdoc comment to this helper function explaining what exactly it's for, maybe providing a basic example of the usage.

**Review by bmish — APPROVED** [review]
This is looking pretty good, thanks for working on it. I think we need a bit more high-level commenting to help people follow what's going on since the logic is a bit complex.

**Review by bmish — CHANGES_REQUESTED** [review]
This is looking pretty good, thanks for working on it. I think we need a bit more high-level commenting, maybe some clearer naming to help people follow what's going on since the logic is a bit complex.

**Comment by VincentMolinie** [comment]
There is an issue on thé Indent. I did a fix at first but rollbacked It cause I didn't know if I should do It in this PR.
This code IS considered valid :
```html
<div>
</div>
  <span>
  </span>
```

The fix of thé rule and thé autofix are quitté easy, let me know if you want It in this PR

File: lib/rules/block-indentation.js (hunk 1 above)
Comment (VincentMolinie) [reply lib/rules/block-indentation.js:24]: That's thé purpose. It says Bad example but It's actually a  good one 😅

File: lib/rules/block-indentation.js (hunk 2 above)
Comment (VincentMolinie) [reply lib/rules/block-indentation.js:185]: Like I said thé indentation of block start IS validated only for thé first node while It should validated all. This code disappear if WE want to really fix this issue

File: lib/rules/block-indentation.js [hunk 5]
```diff
@@ -180,56 +219,132 @@ module.exports = class BlockIndentation extends Rule {
...
+      }
+
+      // Process all the child with the path to access the child from the current node
+      for (const { path: childPath, child } of childrenWithPath) {
+        if (['BlockStatement', 'ElementNode'].includes(child.type)) {
+          this.process(child, childPath);
+        }
+      }
+
+      // Process all the child with the path to access the child from the current node
```
Comment (bmish) [inline lib/rules/block-indentation.js:279]: Duplicate comment from above?

File: test/unit/rules/block-indentation-test.js [hunk 6]
```diff
@@ -227,21 +227,36 @@ generateRuleTests({
         '{{/if}}',
       ].join('\n'),
     },
-  ],
+  ].filter((value, index) => index >= 0),
```
Comment (bmish) [inline test/unit/rules/block-indentation-test.js:230]: Why is this here?

**Comment by bmish** [comment]
> There is an issue on thé Indent. I did a fix at first but rollbacked It cause I didn't know if I should do It in this PR. This code IS considered valid :
> 
> ```
> <div>
> </div>
>   <span>
>   </span>
> ```
> 
> The fix of thé rule and thé autofix are quitté easy, let me know if you want It in this PR

@VincentMolinie it should be fine for us to fix the bug as a follow-up PR. Thanks for working on this!

File: test/unit/rules/block-indentation-test.js (hunk 6 above)
Comment (VincentMolinie) [reply test/unit/rules/block-indentation-test.js:230]: Oops 😂

File: lib/rules/block-indentation.js (hunk 5 above)
Comment (VincentMolinie) [reply lib/rules/block-indentation.js:279]: Good catch 👍 \

**Comment by VincentMolinie** [comment]
> 0e20e0c

Okey so you want me to rollback: 0e20e0c ?

**Comment by VincentMolinie** [comment]
> > There is an issue on thé Indent. I did a fix at first but rollbacked It cause I didn't know if I should do It in this PR. This code IS considered valid :
> > ```
> > <div>
> > </div>
> >   <span>
> >   </span>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The fix of thé rule and thé autofix are quitté easy, let me know if you want It in this PR
> 
> @VincentMolinie it should be fine for us to fix the bug as a follow-up PR. Thanks for working on this!

Okey so you want me to rollback: 0e20e0c ?

**Comment by bmish** [comment]
@VincentMolinie would it be easy for you to extract that bug fix to a follow-up PR after I merge this soon? If so, let's do that quickly. If too inconvenient, we can keep them together.

**Comment by VincentMolinie** [comment]
Done ✅  @bmish

**Comment by bmish** [comment]
@VincentMolinie thanks, did you see the above comment about the bug extraction? Just wondering what you think.

**Comment by VincentMolinie** [comment]
Yes I removed the commit 😉

**Review by bmish — APPROVED** [review]
Thank you!

**Comment by bmish** [comment]
@VincentMolinie excellent, can you create that other PR now?

**Comment by VincentMolinie** [comment]
It's right here: https://github.com/ember-template-lint/ember-template-lint/pull/2318

**Pull request merged**